### ISSUE TITLE
Mempool iterator reset in `broadcastTxRoutine` leads to transaction being resent

### ISSUE THREAD

**Issue by lasarojc** [cometbft/cometbft#1147]
The iterator is reset if the `next` element is `null` in the [routine](https://github.com/cometbft/cometbft/blob/fe45483be36ebfea7e172a3ad949e8fe09a8fd95/mempool/v0/reactor.go#L209). This will possibly cause the same transaction in the mempool to be resent to a given peer.

There are two solutions for this problem:
- Prevent the mempool implementation from resetting iterators to the beginning of the list
- Store in the senders map the peers to which a node sends a transaction, which prevents sending the same transaction to a peer multiple times even when the mempool iterator is reset

Before we opt for one or the other, we would like to discuss the history of the code and the implications of changes to figure why is the code the way it is right now, if it is a deficiency in the iterator code or was it added to solve a bug.

**Comment by cason** [comment]
I have updated the description of the issue, in particular replacing the link to the implementation by one from the `main` branch.

In addition to this investigation, it would be interesting to identify how often this corner case condition is triggered, both in testnets and production networks. The more often it happens, the more relevant it would be the benefit of implementing the second solution.

**Comment by lasarojc** [comment]
Agreed. This will be tested in the fast prototyping but we may want to run it on a real node as well.

**Comment by hvanz** [comment]
The problem is that the iterator in `broadcastTxRoutine` works under the assumption that the transactions are only removed from the beginning of the list, as it happens when creating a block. The variable `next` points to tx just removed to form a block, so `next` will then point the rest of the list, which was not yet processed up to this moment. But tx can be removed in two places: 
- When creating a block and updating the mempool. `ReapMaxBytesMaxGas` returns txs from the beginning of the list to create the block, and `CListMempool.Update` removes them.
- While rechecking txs that the app deems invalid. `resCbRecheck` can remove txs from anywhere on the list. This will cause `next` to point to the beginning, with txs that were already sent.
